net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+		.nte_time       = m0_time_now(),
+		.nte_ep         = listen,
+		.nte_tm         = tm->ftm_ntm,
+	});
+}
+
+/**
+ * Finds queued buffers that timed out and completes them with a
+ * ETIMEDOUT error.
+ */
+static void libfab_tm_buf_timeout(struct m0_fab__tm *ftm)
+{
+	struct m0_net_transfer_mc *net = ftm->ftm_ntm;
+	int                        i;
+	m0_time_t                  now = m0_time_now();
+
+	M0_PRE(libfab_tm_is_locked(ftm));
+	M0_PRE(libfab_tm_invariant(ftm));
+
+	for (i = 0; i < ARRAY_SIZE(net->ntm_q); ++i) {
+		struct m0_net_buffer *nb;
+
+		m0_tl_for(m0_net_tm, &net->ntm_q[i], nb) {
+			if (nb->nb_timeout < now) {
+				nb->nb_flags |= M0_NET_BUF_TIMED_OUT;
+				libfab_buf_done(nb->nb_xprt_private,
+						-ETIMEDOUT);
+			}
+		} m0_tl_endfor;
+	}
+	M0_POST(libfab_tm_invariant(ftm));
+}
+
+/**
+ * Finds pending buffers completions and completes them.
+ *
+ * A buffer is placed on tm::ftm_done queue when its operation is done, but the
+ * completion call-back cannot be immediately invoked, for example, because
+ * completion happened in a synchronous context.
+ */
+static void libfab_tm_buf_done(struct m0_fab__tm *ftm)
+{
+	struct m0_fab__buf *buffer;
+	int                 nr = 0;
+
+	M0_PRE(libfab_tm_is_locked(ftm) && libfab_tm_invariant(ftm));
+	m0_tl_for(fab_buf, &ftm->ftm_done, buffer) {
+		fab_buf_tlist_del(buffer);
+		libfab_buf_complete(buffer, buffer->fb_status);
+		nr++;
+	} m0_tl_endfor;
+
+	if (nr > 0 && ftm->ftm_ntm->ntm_callback_counter == 0)
+		m0_chan_broadcast(&ftm->ftm_ntm->ntm_chan);
+	M0_POST(libfab_tm_invariant(ftm));
+}
+
+/**
+ * Used to handle incoming connection request events
+ */
+static uint32_t libfab_handle_connect_request_events(struct m0_fab__tm *tm)
+{
+	struct m0_fab__ep        *ep = NULL;
+	struct m0_fab__conn_data *cd;
+	struct m0_fab__ep_name    en;
+	struct fid_eq            *eq;
+	struct fi_eq_err_entry    eq_err;
+	struct fi_eq_cm_entry     entry;
+	uint32_t                  event;
+	int                       rc;
+
+	eq = tm->ftm_pep->fep_listen->pep_res.fpr_eq;
+	rc = fi_eq_read(eq, &event, &entry,
+			(sizeof(entry) + sizeof(struct m0_fab__conn_data)), 0);
+	if (rc > 0 && rc >= sizeof(entry)) {
+		if (event == FI_CONNREQ) {
+			memset(&en, 0, sizeof(en));
+			cd = (struct m0_fab__conn_data*)entry.data;
+			libfab_ep_ntop(cd->fcd_netaddr, &en);
+			libfab_fab_ep_find(tm, &en, cd->fcd_straddr, &ep);
+			rc = libfab_conn_accept(ep, tm, entry.info);
+			if (rc != FI_SUCCESS)
+				M0_LOG(M0_ERROR, "Conn accept failed %d", rc);
+			fi_freeinfo(entry.info);
+		}
+	} else if (rc == -FI_EAVAIL) {
+		memset(&eq_err, 0, sizeof(eq_err));
+		rc = fi_eq_readerr(eq, &eq_err, 0);
+		if (rc != sizeof(eq_err)) {
+			M0_LOG(M0_ERROR, "fi_eq_readerr returns error =%s",
+			       fi_strerror((int) -(rc)));
+		} else {
+			M0_LOG(M0_ERROR, "fi_eq_readerr provider err no %d:%s",
+				eq_err.prov_errno,
+				fi_eq_strerror(eq, eq_err.prov_errno,
+					       eq_err.err_data, NULL, 0));
+		}
+	}
```

Comment:
What about other error cases?
e.g. rc < sizeof(entry) && rc != -FI_EAVAIL

Reply:
For all such events, there is no error info available. Hence, all such events can be ignored.

Reply:
I prefer to have comments added this.